### PR TITLE
[Serve] [Docs] Update "Getting Started" documentation

### PR DESCRIPTION
**Pull request by shrekris-anyscale**
## Why are these changes needed?

Recently, Serve has introduced new, fundamental concepts such as deployment graphs. This change updates the "Getting Started" page ([Ray v1.13 link](https://docs.ray.io/en/releases-1.13.0/serve/end_to_end_tutorial.html)) with some of these new concepts.

This change's "Getting Started" page: https://ray--26745.org.readthedocs.build/en/26745/serve/getting_started.html

## Related issue number

N/A

## Checks

- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [X] Unit tests
     - The new code snippets are tested by adding them to the `doc_code` directory.

### CODE REVIEW

File: doc/source/serve/doc_code/getting_started/fastapi_model.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        model_output = self.model(text)
+
+        # Post-process output to return only the translation text
+        translation = model_output[0]["translation_text"]
+
+        return translation
+
+
+ray.init(address="auto")
+serve.start(detached=True)
```
Comment (edoakes) [inline doc/source/serve/doc_code/getting_started/fastapi_model.py:47]: we do not need `serve.start` anymore when using `serve.run (indeed it is deprecated). please update relevant code samples

Comment (edoakes) [reply doc/source/serve/doc_code/getting_started/fastapi_model.py:47]: you can also drop the `ray.init` call

File: doc/source/serve/getting_started.md [hunk 2]
```diff
@@ -66,43 +61,50 @@ PyTorch, and Tensorflow for more info and examples:
...
-instance and deploying to a Ray cluster.
+The `Translator` class has two modifications:
+1. It has a decorator, `@serve.deployment`.
+2. It has a new method, `__call__`.
+
+The decorator converts `Translator` from a Python class into a Ray Serve
+`Deployment` object.
+
+Each deployment stores a single Python function or class that you write and uses
+it to serve requests. You can scale and configure each of your deployments using
```
Comment (edoakes) [inline doc/source/serve/getting_started.md:91]:
```suggestion
it to serve requests. You can scale and configure each of your deployments independently using
```

just to drive it home

File: doc/source/serve/getting_started.md [hunk 3]
```diff
@@ -66,43 +61,50 @@ PyTorch, and Tensorflow for more info and examples:
...
+parameters in the `@serve.deployment` decorator.
+
+Deployments receive Starlette HTTP `request` objects [^f1]. If your deployment stores a Python function, the function is called on this `request` object. If you deployment stores a class, the class's `__call__` method is called on this `request` object. The return value is sent back in the HTTP response body.
+
+This is why `Translator` needs a new `__call__` method. The method processes the incoming HTTP request by reading its JSON data and forwarding it to the `translate` method. The translated text is returned and sent back through the HTTP response.
```
Comment (edoakes) [inline doc/source/serve/getting_started.md:96]: here we should mention that you can also use fastapi to not use raw http requests with a pointer to that page

File: doc/source/serve/getting_started.md [hunk 4]
```diff
@@ -126,90 +128,50 @@ script exits. If you would rather stop Ray Serve after the script exits, use
...
-into `@serve.deployment`, the `request` uses the function's name by
-default. For example, if the decorator is `@serve.deployment` and the
-function's signature is `def manager(request)`, the HTTP request should use
-`manager`.
-:::
+`translator = Translator.bind()` binds our deployment to constructor `args` and
+`kwargs`. Ray Serve uses these to initialize the `Translator` class using its
+`__init__` method before it starts serving traffic. The `translator` variable
+is a `DeploymentNode` object. As we'll see later in the tutorial, you can also
+connect `DeploymentNodes` together to form what's called a `Deployment Graph`.
```
Comment (edoakes) [inline doc/source/serve/getting_started.md:144]:
```suggestion
connect `DeploymentNodes` together to form what's called a `Deployment Graph` to serve multiple models instead of just one.
```

**Review by edoakes — COMMENTED** [review]
Should we separate defining the call graph and the HTTP adapters as separate sections?

I'm also not entirely sure if we should include these in "Getting Started," it makes it a bit long and complicated. I think we want this to focus on the very core functionality, then branch out into more complex usage.

I would actually suggest that in this section we only introduce `serve.run`ing a single deployment, then binding and `serve.run`ing multiple together. In a separate section we can introduce the call graph building.

Thoughts @shrekris-anyscale @simon-mo @sihanwang41?

**Comment by shrekris-anyscale** [comment]
> I would actually suggest that in this section we only introduce `serve.run`ing a single deployment, then binding and `serve.run`ing multiple together. In a separate section we can introduce the call graph building.

@edoakes I think that makes sense, but how would we introduce `serve.run` for multiple deployments without creating a call graph?

> Should we separate defining the call graph and the HTTP adapters as separate sections?

I also sensed that this complicated the tutorial a bit. I included them because I felt that the HTTP adapters are meant to make it easier to get started with Serve, and I couldn't figure out how to discuss the adapters without discussing the call graph. Do you think it would be better to just omit both here?

File: doc/source/serve/doc_code/getting_started/fastapi_model.py (hunk 1 above)
Comment (shrekris-anyscale) [reply doc/source/serve/doc_code/getting_started/fastapi_model.py]: I dropped `serve.start()`, but I kept `ray.init()` for now. I think it's useful to have users explicitly start a Ray cluster and connect to it in the tutorial to help them build a clearer mental model of how Ray Serve interacts with Ray. What do you think?

File: doc/source/serve/getting_started.md (hunk 3 above)
Comment (shrekris-anyscale) [reply doc/source/serve/getting_started.md]: Good idea, I added a link.

**Comment by shrekris-anyscale** [comment]
Ah my bad, did a bad rebase.

**Review by simon-mo — COMMENTED** [review]
- let's drop all `ray.init` and `ray start`; favoring `serve.run` and `serve run`
- "Deployment Graphs and HTTP Adapters"'s wall of text is a little bit hard to parse. Can you potentially add line numbers here?

File: doc/source/serve/getting_started.md [hunk 5]
```diff
@@ -2,9 +2,20 @@
...
 # Getting Started
 
-This tutorial will walk you through the process of using Ray Serve to deploy a single model behind HTTP locally.
+This tutorial will walk you through the process of deploying models with Ray Serve. It will show you how to
 
-We'll be using [HuggingFace's SummarizationPipeline](https://huggingface.co/docs/transformers/main_classes/pipelines#transformers.SummarizationPipeline) to deploy a model that summarizes text.
+* Expose your models over HTTP using Ray Serve `deployments`
+* Scale your deployments to meet your workload's requirements
+* Allocate resources like fractional GPUs and CPUs to your deployments
+* Compose multiple-model machine learning pipelines with Ray Serve `deployment graphs`
```
Comment (architkulkarni) [inline doc/source/serve/getting_started.md:10]: Nit: monospace looks a little odd for `deployments` and `deployment graphs`.  Is there a deployment graph page we can link to here?

File: doc/source/serve/getting_started.md [hunk 6]
```diff
@@ -14,48 +25,43 @@ If you have suggestions on how to improve this tutorial,
 To run this example, you will need to install the following:
 
 ```bash
-$ pip install "ray[serve]" transformers
+$ pip install "ray[serve]" transformers requests
```
Comment (architkulkarni) [inline doc/source/serve/getting_started.md:28]:
```suggestion
pip install "ray[serve]" transformers requests
```
nit, makes it copypastable

File: doc/source/serve/getting_started.md [hunk 7]
```diff
@@ -14,48 +25,43 @@ If you have suggestions on how to improve this tutorial,
...
 ```
 
 
 ## Example Model
 
-Let's first take a look at how the model works without using Ray Serve.
-This is the code for the model:
+First, let's take a look at our text-translation model. Here's it's code:
```
Comment (architkulkarni) [inline doc/source/serve/getting_started.md:34]:
```suggestion
First, let's take a look at our text-translation model. Here's its code:
```

File: doc/source/serve/getting_started.md [hunk 8]
```diff
@@ -14,48 +25,43 @@ If you have suggestions on how to improve this tutorial,
...
+:start-after: __start_translation_model__
+:end-before: __end_translation_model__
 :language: python
 :linenos: true
-:start-after: __local_model_start__
 ```
 
-The Python file, called `local_model.py` uses the `summarize` function to
-generate summaries of text.
+The Python file, called `model.py` uses the `Translator` class to translate English text to French.
```
Comment (architkulkarni) [inline doc/source/serve/getting_started.md:43]:
```suggestion
The Python file, called `model.py`, uses the `Translator` class to translate English text to French.
```

File: doc/source/serve/getting_started.md [hunk 9]
```diff
@@ -65,339 +71,279 @@ PyTorch, and Tensorflow for more info and examples:
...
+
+This command will start running `Translator` and then block. It can be killed with `ctrl-C` in the terminal.
 
-Since `@serve.deployment` makes `router` a `Deployment` object, it can be
-deployed using `router.deploy()`:
+## Testing Ray Serve Deployments
 
-```{literalinclude} ../../../python/ray/serve/examples/doc/e2e_deployment.py
-:end-before: __router_deploy_end__
+We can now test our model over HTTP. It can be reached at `http://127.0.0.1:8000/`
```
Comment (architkulkarni) [inline doc/source/serve/getting_started.md:138]: Missing a period but looks awkward to add a period after a URL.  Maybe:

It can be reached at the following URL:

```
http://127.0.0.1:800
```

File: doc/source/serve/getting_started.md [hunk 10]
```diff
@@ -65,339 +71,279 @@ PyTorch, and Tensorflow for more info and examples:
...
-[query parameter](https://fastapi.tiangolo.com/tutorial/query-params/)
-feature.
+c'était le meilleur des temps, c'était le pire des temps .
+```
 
-Since we still deploy our model locally, the full URL still uses the
-localhost IP. This means each of our three routes comes after the
-`http://127.0.0.1:8000` IP and port address. As an example, we can make
-requests to the `max10` route using this client script:
+Deployment graphs are useful since they let you deploy each part of your Machine Learning pipeline, such as inference and business logic steps, in separate deployments. Each of these deployments can be individually configured and scaled, ensuring you get maximal performance from your resources.
```
Comment (architkulkarni) [inline doc/source/serve/getting_started.md:304]:
```suggestion
Deployment graphs are useful since they let you deploy each part of your machine learning pipeline, such as inference and business logic steps, in separate deployments. Each of these deployments can be individually configured and scaled, ensuring you get maximal performance from your resources.
```

File: doc/source/serve/getting_started.md [hunk 11]
```diff
@@ -14,48 +25,43 @@ If you have suggestions on how to improve this tutorial,
 To run this example, you will need to install the following:
 
 ```bash
-$ pip install "ray[serve]" transformers
+$ pip install "ray[serve]" transformers requests
 ```
 
 
 ## Example Model
```
Comment (architkulkarni) [inline doc/source/serve/getting_started.md:32]: Maybe rename this to "Example Model: Before Ray Serve" or similar?  As a new reader, I first glance at the first big code "Example" and it throws me off slightly that it doesn't have any Ray Serve code.

File: doc/source/serve/getting_started.md [hunk 12]
```diff
@@ -14,48 +25,43 @@ If you have suggestions on how to improve this tutorial,
...
-This is the code for the model:
+First, let's take a look at our text-translation model. Here's it's code:
 
-```{literalinclude} ../../../python/ray/serve/examples/doc/e2e_local.py
-:end-before: __local_model_end__
+```{literalinclude} ../serve/doc_code/getting_started/models.py
+:start-after: __start_translation_model__
+:end-before: __end_translation_model__
 :language: python
 :linenos: true
```
Comment (architkulkarni) [inline doc/source/serve/getting_started.md:40]: Is there an easy way to make the line numbers not get copied when using this button?

Screen Shot 2022-07-29 at 2 36 43 PM

**Comment by architkulkarni** [comment]
Looks great, just some minor questions/suggestions.

As for line numbers, one alternative I've seen is to omit line numbers altogether and add little reference numbers in the comments so that you can refer to them in the text, e.g. https://github.com/fluentpython/example-code-2e/blob/master/07-1class-func/bingocall.py. But it's probably better to be consistent with whatever we're doing elsewhere in the docs.

File: doc/source/serve/getting_started.md (hunk 5 above)
Comment (shrekris-anyscale) [reply doc/source/serve/getting_started.md:10]: I linked them both to their corresponding documentation instead of using monospace.

File: doc/source/serve/getting_started.md (hunk 9 above)
Comment (shrekris-anyscale) [reply doc/source/serve/getting_started.md:138]: I replaced the sentence with your suggestion.

File: doc/source/serve/getting_started.md (hunk 11 above)
Comment (shrekris-anyscale) [reply doc/source/serve/getting_started.md:32]: I changed it to "Model Example: Before Ray Serve".

File: doc/source/serve/getting_started.md (hunk 12 above)
Comment (shrekris-anyscale) [reply doc/source/serve/getting_started.md:40]: Unfortunately not– this is a known issue: https://github.com/ray-project/ray/issues/21968.

Users can highlight and copy the text without copying numbers, but the clipboard will pick up the numbers. For consistency, I'd prefer to keep these numbers instead of using commented references, but I'm open to changing them.

Comment (architkulkarni) [reply doc/source/serve/getting_started.md:40]: Cool, keeping it this way sounds good

**Review by architkulkarni — APPROVED** [review]

File: doc/source/serve/getting_started.md [hunk 13]
```diff
@@ -65,339 +71,283 @@ PyTorch, and Tensorflow for more info and examples:
...
+it to serve requests. You can scale and configure each of your deployments independently using
+parameters in the `@serve.deployment` decorator.
 
-:::{note}
-`ray.init()` connects to or starts a single-node Ray cluster on your
-local machine,  which allows you to use all your CPU cores to serve
-requests in parallel. To start a multi-node cluster, see
-{ref}`serve-deploy-tutorial`.
-:::
+Deployments receive Starlette HTTP `request` objects [^f1]. If your deployment stores a Python function, the function is called on this `request` object. If you deployment stores a class, the class's `__call__` method is called on this `request` object. The return value is sent back in the HTTP response body.
```
Comment (sihanwang41) [inline doc/source/serve/getting_started.md:105]: Can we put the explanation like **note** section?

like
image


And then user can easily find  ```Next, we need to...```

File: doc/source/serve/getting_started.md [hunk 14]
```diff
@@ -2,9 +2,20 @@
 
 # Getting Started
```
Comment (sihanwang41) [inline doc/source/serve/getting_started.md:3]: Can we have a table of contents in this page? Lets user target their interests quickly.

**Review by simon-mo — APPROVED** [review]
